_posts/2017-07-04-flink-rescalable-state.md

```diff
@@ -0,0 +1,172 @@
+
+You can see detail about Flink’s checkpoints in [the documentation](https://ci.apache.org/projects/flink/flink-docs-release-1.3/internals/stream_checkpointing.html). In a nutshell, a checkpoint is triggered when a checkpoint coordinator injects a special event (a so-called checkpoint barrier) into a stream.
+
+Checkpoint barriers flow downstream with the event stream from sources to sinks, and whenever an operator instance receives a barrier, the operator instance immediately snapshots its current state to a distributed storage system, e.g. HDFS.
+
+On restore, the new tasks for the job (which potentially run on different machines now) can again pick up the state data from the distributed storage system.
+
+<br><center><i>Figure 1</i></center>
+<center>
+<img src="{{ site.baseurl }}/img/blog/stateless-stateful-streaming.svg" style="width:70%;margin:10px">
+</center>
+<br>
+
+We can piggyback rescaling of stateful jobs on checkpointing, as shown in **Figure 1B**. First, a checkpoint is triggered and sent to a distributed storage system. Next, the job is restarted with a changed parallelism and can access a consistent snapshot of all previous state from the distributed storage. While this solves (i) redistribution of a (ii) consistent state across machines there is still one problem: without a clear 1:1 relationship between previous state and new parallel operator instances, how can we assign the state in a (iii) meaningful way?
+
+We could again assign the state from previous `map_1` and `map_2` to the new `map_1` and `map_2`. But this would leave `map_3` with empty state. Depending on the type of state and concrete semantics of the job, this naive approach could lead to anything from inefficiency to incorrect results.
+
+In the following section, we’ll explain how we solved the problem of efficient, meaningful state reassignment in Flink. Each of Flink state’s two flavours, operator state and keyed state, requires a different approach to state assignment.
+
+## Reassigning Operator State When Rescaling
+
+First, we’ll discuss how state reassignment in rescaling works for operator state. A common real-world use-case of operator state in Flink is to maintain the current offsets for Kafka partitions in Kafka sources. Each Kafka source instance would maintain `<PartitionID, Offset>` pairs–one pair for each Kafka partition that the source is reading–as operator state. How would we redistribute this operator state in case of rescaling? Ideally, we would like to reassign all `<PartitionID, Offset>` pairs from the checkpoint in round robin across all parallel operator instances after the rescaling.
+
+As a user, we are aware of the “meaning” of Kafka partition offsets, and we know that we can treat them as independent, redistributable units of state. The problem remains how we can we share this domain-specific knowledge with Flink.
```

Comment:
"The problem remains how we can we share this domain-specific knowledge with Flink."
maybe:
"The problem of how we can share this domain-specific knowledge with Flink remains."?